proxy/srs.go

```diff
@@ -0,0 +1,553 @@
+}
+
+// SRSLoadBalancer is the interface to load balance the SRS servers.
+type SRSLoadBalancer interface {
+	// Initialize the load balancer.
+	Initialize(ctx context.Context) error
+	// Update the backer server.
+	Update(ctx context.Context, server *SRSServer) error
+	// Pick a backend server for the specified stream URL.
+	Pick(ctx context.Context, streamURL string) (*SRSServer, error)
+	// Load or store the HLS streaming for the specified stream URL.
+	LoadOrStoreHLS(ctx context.Context, streamURL string, value *HLSPlayStream) (*HLSPlayStream, error)
+	// Load the HLS streaming by SPBHID, the SRS Proxy Backend HLS ID.
+	LoadHLSBySPBHID(ctx context.Context, spbhid string) (*HLSPlayStream, error)
+	// Store the WebRTC streaming for the specified stream URL.
+	StoreWebRTC(ctx context.Context, streamURL string, value *RTCConnection) error
+	// Load the WebRTC streaming by ufrag, the ICE username.
+	LoadWebRTCByUfrag(ctx context.Context, ufrag string) (*RTCConnection, error)
+}
+
+// srsLoadBalancer is the global SRS load balancer.
+var srsLoadBalancer SRSLoadBalancer
+
+// srsMemoryLoadBalancer stores state in memory.
+type srsMemoryLoadBalancer struct {
+	// All available SRS servers, key is server ID.
+	servers sync.Map[string, *SRSServer]
+	// The picked server to servce client by specified stream URL, key is stream url.
+	picked sync.Map[string, *SRSServer]
+	// The HLS streaming, key is stream URL.
+	hlsStreamURL sync.Map[string, *HLSPlayStream]
+	// The HLS streaming, key is SPBHID.
+	hlsSPBHID sync.Map[string, *HLSPlayStream]
+	// The WebRTC streaming, key is stream URL.
+	rtcStreamURL sync.Map[string, *RTCConnection]
+	// The WebRTC streaming, key is ufrag.
+	rtcUfrag sync.Map[string, *RTCConnection]
+}
+
+func NewMemoryLoadBalancer() SRSLoadBalancer {
+	return &srsMemoryLoadBalancer{}
+}
+
+func (v *srsMemoryLoadBalancer) Initialize(ctx context.Context) error {
+	if server, err := NewDefaultSRSForDebugging(); err != nil {
+		return errors.Wrapf(err, "initialize default SRS")
+	} else if server != nil {
+		if err := v.Update(ctx, server); err != nil {
+			return errors.Wrapf(err, "update default SRS %+v", server)
+		}
+
+		// Keep alive.
+		go func() {
+			for {
+				select {
+				case <-ctx.Done():
+					return
+				case <-time.After(30 * time.Second):
+					if err := v.Update(ctx, server); err != nil {
+						logger.Wf(ctx, "update default SRS %+v failed, %+v", server, err)
+					}
+				}
+			}
+		}()
+		logger.Df(ctx, "MemoryLB: Initialize default SRS media server, %+v", server)
+	}
+	return nil
+}
+
+func (v *srsMemoryLoadBalancer) Update(ctx context.Context, server *SRSServer) error {
+	v.servers.Store(server.ID(), server)
+	return nil
+}
+
+func (v *srsMemoryLoadBalancer) Pick(ctx context.Context, streamURL string) (*SRSServer, error) {
+	// Always proxy to the same server for the same stream URL.
+	if server, ok := v.picked.Load(streamURL); ok {
+		return server, nil
+	}
+
+	// Gather all servers, alive in 60s ago.
```

Comment:
It's gather server within `srsServerAliveDuration = 300s` ago.

Reply:
Fixed.